src/H5Dselect.c

```diff
@@ -256,6 +256,187 @@ H5D__select_io(const H5D_io_info_t *io_info, size_t elmt_size, size_t nelmts, H5
+        if (NULL == (dst_sel_iter = H5FL_MALLOC(H5S_sel_iter_t)))
+            HGOTO_ERROR(H5E_DATASPACE, H5E_CANTALLOC, FAIL, "can't allocate destination selection iterator")
+        if (NULL == (src_sel_iter = H5FL_MALLOC(H5S_sel_iter_t)))
+            HGOTO_ERROR(H5E_DATASPACE, H5E_CANTALLOC, FAIL, "can't allocate source selection iterator")
+
+        /* Initialize destination selection iterator */
+        if (H5S_select_iter_init(dst_sel_iter, dst_space, elmt_size, sel_iter_flags) < 0)
+            HGOTO_ERROR(H5E_DATASPACE, H5E_CANTINIT, FAIL, "unable to initialize selection iterator")
+        dst_sel_iter_init = TRUE; /* Destination selection iteration info has been initialized */
+
+        /* Initialize source selection iterator */
+        if (H5S_select_iter_init(src_sel_iter, src_space, elmt_size, H5S_SEL_ITER_SHARE_WITH_DATASPACE) < 0)
+            HGOTO_ERROR(H5E_DATASPACE, H5E_CANTINIT, FAIL, "unable to initialize selection iterator")
+        src_sel_iter_init = TRUE; /* Source selection iteration info has been initialized */
+
+        /* Initialize sequence counts */
+        curr_dst_seq = curr_src_seq = 0;
+        dst_nseq = src_nseq = 0;
+
+        /* Loop, until all bytes are processed */
+        while (nelmts > 0) {
+            /* Check if more destination buffer sequences are needed */
+            if (curr_dst_seq >= dst_nseq) {
+                /* Get sequences for destination selection */
+                if (H5S_SELECT_ITER_GET_SEQ_LIST(dst_sel_iter, vec_size, nelmts, &dst_nseq, &dst_nelem,
+                                                 dst_off, dst_len) < 0)
+                    HGOTO_ERROR(H5E_DATASPACE, H5E_CANTGET, FAIL, "sequence length generation failed")
+
+                /* Start at the beginning of the sequences again */
+                curr_dst_seq = 0;
+            }
+
+            /* Check if more source buffer sequences are needed */
+            if (curr_src_seq >= src_nseq) {
+                /* Get sequences for source selection */
+                if (H5S_SELECT_ITER_GET_SEQ_LIST(src_sel_iter, vec_size, nelmts, &src_nseq, &src_nelem,
+                                                 src_off, src_len) < 0)
+                    HGOTO_ERROR(H5E_DATASPACE, H5E_CANTGET, FAIL, "sequence length generation failed")
+
+                /* Start at the beginning of the sequences again */
+                curr_src_seq = 0;
+            } /* end if */
+
+            /* Perform vectorized memcpy from src_buf to dst_buf */
+            if ((bytes_copied = H5VM_memcpyvv(dst_buf, dst_nseq, &curr_dst_seq, dst_len, dst_off, src_buf,
+                                              src_nseq, &curr_src_seq, src_len, src_off)) < 0)
+                HGOTO_ERROR(H5E_IO, H5E_WRITEERROR, FAIL, "vectorized memcpy failed")
+
+            /* Decrement number of elements left to process */
+            HDassert(((size_t)bytes_copied % elmt_size) == 0);
+            if (elmt_size == 0)
```

Comment:
Shouldn't this just be an assert at the top of the function?

Reply:
Hmm, this was copied from H5D__select_io, but yes, I think in both cases the error check should be moved to the top of the functions since elmt_size shouldn't change.